testament/testament.nim

```diff
@@ -851,4 +884,437 @@ proc main() =
+    dryRun,            ## do not run the tests, only indicate which would run
+    rerunFailed,       ## only run tests failed in the previous run
+    runKnownIssues     ## also execute tests marked as known issues
+
+  ExecutionFlags = set[ExecutionFlag]
+    ## track the option flags that got set for an execution
+
+  RetryList = OrderedTable[TestId, RetryInfo]
+      ## record failures in here so the user can choose to retry them
+
+  TestTargets = set[TestTarget]
+
+  DebugInfo = OrderedTable[ActionId, string]
+
+  RunTime = object
+    compileStart: float      ## when the compile process start
+    compileEnd: float        ## when the compile process ends
+    checkStart: float        ## for run or compiles, check output start
+    checkEnd: float          ## for run or compiles, check output end
+    runStart: float          ## for run, start of execution
+    runEnd: float            ## for run, end of execution
+
+  TestRun = object
+    testId: TestId           ## test id for which this belongs
+    target: TestTarget       ## which target to run for
+    matrixEntry: EntryId     ## which item from the matrix was used
+    runtime: RunTime         ## time tracking for test activities
+
+  TestAction = object
+    runId: RunId
+    case kind: TTestAction
+    of actionReject, actionRun:
+      discard
+    of actionCompile:
+      partOfRun: bool
+
+  Execution = object
+    # user and execution inputs
+    filter: TestFilter       ## filter that was configured
+    flags: ExecutionFlags    ## various options set by the user
+    targets: TestTargets     ## specified targets or `noTargetsSpecified`
+    workingDir: string       ## working directory to begin execution in
+    nimSpecified: bool       ## whether the user specified the nim
+    testArgs: string         ## arguments passed to tests by the user
+
+    # environment input / setup
+    compilerPath: string     ## compiler command to use
+    testsDir: string         ## where to look for tests
+
+    # test discovery data
+    testCats:  Categories    ## categories discovered, for this execution
+    testFiles: seq[TestFile] ## files for this execution
+    testSpecs: seq[TSpec]    ## spec for each file
+
+    # test execution data
+    testRuns: seq[TestRun]   ## a test run: reject, compile, or compile + run
+    actions: seq[TestAction] ## test actions for each run, phases of a run
+    debugInfo: DebugInfo     ## debug info related to actions run for tests
+
+    # test execution related data
+    retryList: RetryList     ## list of failures to potentially retry later
+
+  ParseCliResult = enum
+    parseSuccess       ## successfully parsed cli params
+    parseQuitWithUsage ## parsing failed, quit with usage message
+
+const
+  testResultsDir = "testresults"
+  cacheResultsDir = testResultsDir / "cacheresults"
+  noTargetsSpecified: TestTargets = {}
+  defaultExecFlags = {outputColour}
+  defaultBatchSize = 10
+  noMatrixEntry: EntryId = -1
```

Comment:
@alaviss as you can see I learned similar lessons with DOD. 😆

Reply:
You really should turn those `int` into `int32`, no one do more than 2M tests yet

Reply:
Done locally.